### DISCUSSION TITLE
How to work with Languages in Dataverse.

### DISCUSSION THREAD

**Discussion by MattB-msft** [microsoft/PowerPlatform-DataverseServiceClient#238] [Show and tell]
In a recent discussion, concerning how to detect language availability and install language packs on dataverse, I realized that we really didn't have a concise location for the various means to work with languages in the dataverse system. 
I ended up writing up a quick what to use for what list for the person asking. 
 
In the effort to head of further confusion about it, I will copy that here: 

In all cases, Dataverse "the system" refers to Languages by LCID (Language Code Identifier).  
Various applications that are built on top of the Dataverse system will use ISO formats for readability.   
For example:  ISO:en-us|LCID:1033

**1.  Getting the current Default language of an environment:**
You can get the default language of the environment by asking the [_organization_](https://docs.microsoft.com/en-us/powerapps/developer/data-platform/reference/entities/organization) table for it. The field is [**_languagecode_**](https://docs.microsoft.com/en-us/powerapps/developer/data-platform/reference/entities/organization#BKMK_LanguageCode)

**2. Getting the current list of Languages that are enabled on a given Dataverse Environment:**
To get all languages active on the environment (meaning installed and ready to us) you would use [RetrieveAvailableLanguagesRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.retrieveavailablelanguagesrequest?view=dynamics-general-ce-9).  This will return a list of LCID’s that are enabled on the org. 

**3. Getting the list of language packs that are available to a given environment:**
If you want to see all languages available to the server, ( not necessary activated) you would use [RetrieveInstalledLanguagePacksRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.retrieveinstalledlanguagepacksrequest?view=dynamics-general-ce-9) – 
_**Note:** 1033 will generally not appear here because it's not a language pack, it’s the native language, unless the server is built using a different language._ 

**4. Getting the version of a language pack installed on a Dataverse environment:**
To get the version of the language pack installed you would use [RetrieveInstalledLanguagePackVersionRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.retrieveinstalledlanguagepackversionrequest.language?view=dynamics-general-ce-9).  Which will give you the version number of the language pack installed.

**5. To issue an Install command for a language pack:**
To install a language pack you would use either [ProvisionLanguageAsyncRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.provisionlanguageasyncrequest?view=dynamics-general-ce-9) or [ProvisionLanguageRequest](https://docs.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.provisionlanguagerequest?view=dynamics-general-ce-9).  The recommendation is to always use the Async version of this request,  it will return an AsyncOperationId which you can then use to monitor status of the install request via query to the asyncoperation table.

**Comment by JimDaly** [comment]
We have this topic that covers working with languages programmatically
[Customize labels to support multiple languages](https://docs.microsoft.com/en-us/powerapps/developer/data-platform/customize-labels-support-multiple-languages)

But it doesn't call out that [Organization.LanguageCode](https://docs.microsoft.com/en-us/powerapps/developer/data-platform/reference/entities/organization#BKMK_LanguageCode) has the base language.
